File: trunk/MapXtreme/6.6/Samples/WebControlsSourceCode/cs/Tools/ZoomBarControl.cs
using System;
using System.Drawing;
using System.IO;
using System.Web.SessionState;
using System.Web.UI;
using System.Web.UI.Design;
using System.Web.UI.WebControls;
using System.ComponentModel;
using System.Xml;
using System.Xml.XPath;
using System.Xml.Xsl;

namespace MapInfo.WebControls
{
	/// <summary>
	/// This tool allows you to zoom to a given fixed level.
	/// </summary>
	/// <remarks>
	/// This tool behaves differently from other tools like zoom-in. When the mouse hovers over the tool or goes outside the tool 
	///  it uses a different image to give users the impression of being active or not active. There is no interaction on the map. A click on the tool zooms the map 
	///  to level provided. The zoom level is in the map's current units.
	///  This control can be part of a zoom bar which allows users to zoom to various pre-set levels. You can drop multiple zoom bar tools and set
	///  different levels for each.
	/// </remarks>
	[
	ToolboxData("<{0}:ZoomBarTool runat=server></{0}:ZoomBarTool>")
	]
	public class ZoomBarTool : WebTool
	{
		/// <summary>
		/// Constructor for the tool.
		/// </summary>
		/// <remarks>
		/// Sets default values for parameters.
		/// </remarks>
		public ZoomBarTool()
		{
			ZoomLevel = 0.0;
			InactiveImageUrl =  string.Format("{0}/{1}Inactive.gif", Resources.ResourceFolder, GetType().Name);
			ActiveImageUrl = string.Format("{0}/{1}Active.gif", Resources.ResourceFolder, GetType().Name);
			Active = false;
			Command = CommandEnum.ZoomToLevel.ToString();
			ClientInteraction = ClientInteractionEnum.NullInteraction.ToString();
			ClientCommand = ClientCommandEnum.ZoomCommand.ToString();
		}

		/// <summary>
		/// This property is not used for this control.
		/// </summary>
		/// <remarks>Since there is no activation or deactivation of the tool, this property is not to be used.</remarks>
		[
		Browsable(false)
		]
		public new bool Active {
			get {
				return (bool)ViewState["Active"];
			}
			set {
				ViewState["Active"] = value;
			}
		}

		/// <summary>
		/// Zoom level which the map should be zoomed.
		/// </summary>
		/// <remarks>The value is provided in the map's exisiting units.</remarks>
		[
		Browsable(true),
		DesignerSerializationVisibility(DesignerSerializationVisibility.Visible),
		PersistenceMode(PersistenceMode.Attribute),
		]
		public double ZoomLevel {
			get {
				return (double)ViewState["ZoomLevel"];
			}
			set {
				ViewState["ZoomLevel"] = value;
			}
		}

		/// <summary>
		/// Renders the control.
		/// </summary>
		/// <remarks>This method renders the image and events for the tool. This method writes javascript code to create a client side command.</remarks>
		/// <param name="writer">HtmlTextWriter</param>
		protected override void RenderContents(HtmlTextWriter writer) {
			if (Page !=null) {
				Page.VerifyRenderingInServerForm(this);
			}

			// Render HTML
			writer.AddAttribute(HtmlTextWriterAttribute.Id, UniqueID + "_Image");
			writer.AddAttribute(HtmlTextWriterAttribute.Src, InactiveImageUrl);
			writer.AddAttribute("OnMouseOver", string.Format("javascript:this.src='{0}';", ActiveImageUrl));
			writer.AddAttribute("OnMouseOut", string.Format("javascript:this.src='{0}';", InactiveImageUrl));
			ToolTip = L10NUtils.Resources.GetString("ZoomBarMapToolHelpText");
			writer.AddAttribute(HtmlTextWriterAttribute.Title, ToolTip);
			writer.AddAttribute(HtmlTextWriterAttribute.Alt, ToolTip);
			writer.RenderBeginTag(HtmlTextWriterTag.Img);
			writer.RenderEndTag();

			writer.AddAttribute("language", "javascript");
			writer.AddAttribute(HtmlTextWriterAttribute.Type, "text/javascript");
			writer.RenderBeginTag(HtmlTextWriterTag.Script);
			writer.WriteLine(string.Format("var {0}Int = new {1}('{2}_Image', null);", UniqueID, ClientInteraction, MapControlID));
			writer.WriteLine(string.Format("var {0}Cmd = new {1}('{2}', {0}Int);", UniqueID, ClientCommand, Command));
			writer.WriteLine(string.Format("var {0}Me = FindElement('{1}_Image');", UniqueID,  UniqueID));
			writer.WriteLine(string.Format("{0}Cmd.zoomLevel = {1};", UniqueID, ZoomLevel.ToString()));
			writer.WriteLine(string.Format("{0}Me.onclick = {0}Cmd.Exc;", UniqueID));
			writer.RenderEndTag();
		}
	}
}

File: trunk/项目开发进度2009年2月9日/SeeMapCell/MapXtreme/6.6/Samples/WebControlsSourceCode/vb/Tools/ZoomBarControl.vb
Imports System
Imports System.Drawing
Imports System.IO
Imports System.Web.SessionState
Imports System.Web.UI
Imports System.Web.UI.Design
Imports System.Web.UI.WebControls
Imports System.ComponentModel
Imports System.Xml
Imports System.Xml.XPath
Imports System.Xml.Xsl

'/ <summary>
'/ This control depends upon mapControl, does rubberbanding drawing on client side. All it knows is the HTML image tag its doing
'/ drawing on.
'/ </summary>
<ToolboxData("<{0}:ZoomBarTool runat=server></{0}:ZoomBarTool>")> _
 Public Class ZoomBarTool
    Inherits WebTool
    Public Sub New()
        ZoomLevel = 0.0
        InactiveImageUrl = String.Format("{0}/{1}Inactive.gif", Resources.ResourceFolder, MyBase.GetType.Name)
        ActiveImageUrl = String.Format("{0}/{1}Active.gif", Resources.ResourceFolder, MyBase.GetType.Name)
        Active = False
        Command = CommandEnum.ZoomToLevel.ToString()
        ClientInteraction = ClientInteractionEnum.NullInteraction.ToString()
        ClientCommand = ClientCommandEnum.ZoomCommand.ToString()
    End Sub

    <Browsable(True), DesignerSerializationVisibility(DesignerSerializationVisibility.Visible), PersistenceMode(PersistenceMode.Attribute)> _
    Public Property ZoomLevel() As Double
        Get
            Return CType(ViewState("ZoomLevel"), Double)
        End Get
        Set(ByVal Value As Double)
            ViewState("ZoomLevel") = Value
        End Set
    End Property
    <Browsable(False)> _
    Public Shadows Property Active() As Boolean
        Get
            Return CType(ViewState("Active"), Boolean)
        End Get
        Set(ByVal Value As Boolean)
            ViewState("Active") = Value
        End Set
    End Property

    '/ <summary>
    '/ Render this control to the output parameter specified.
    '/ </summary>
    '/ <param name="writer"> The HTML writer to write out to </param>
    Protected Overrides Sub RenderContents(ByVal writer As HtmlTextWriter)

        If Not Page Is Nothing Then
            Page.VerifyRenderingInServerForm(Me)
        End If

        ' Render HTML
        writer.AddAttribute(HtmlTextWriterAttribute.Id, UniqueID + "_Image")
        writer.AddAttribute(HtmlTextWriterAttribute.Src, InactiveImageUrl)
        writer.AddAttribute("OnMouseOver", String.Format("javascript:this.src='{0}';", ActiveImageUrl))
        writer.AddAttribute("OnMouseOut", String.Format("javascript:this.src='{0}';", InactiveImageUrl))
        ToolTip = L10NUtils.Resources.GetString("ZoomBarMapToolHelpText")
        writer.AddAttribute(HtmlTextWriterAttribute.Title, ToolTip)
        writer.AddAttribute(HtmlTextWriterAttribute.Alt, ToolTip)
        writer.RenderBeginTag(HtmlTextWriterTag.Img)
        writer.RenderEndTag()

        writer.AddAttribute("language", "javascript")
        writer.AddAttribute(HtmlTextWriterAttribute.Type, "text/javascript")
        writer.RenderBeginTag(HtmlTextWriterTag.Script)
        writer.WriteLine(String.Format("var {0}Int = new {1}('{2}_Image', null);", UniqueID, ClientInteraction, MapControlID))
        writer.WriteLine(String.Format("var {0}Cmd = new {1}('{2}', {0}Int);", UniqueID, ClientCommand, Command))
        writer.WriteLine(String.Format("var {0}Me = FindElement('{1}_Image');", UniqueID, UniqueID))
        writer.WriteLine(String.Format("{0}Cmd.zoomLevel = {1};", UniqueID, ZoomLevel.ToString()))
        writer.WriteLine(String.Format("{0}Me.onclick = {0}Cmd.Exc;", UniqueID))
        writer.RenderEndTag()
    End Sub
End Class

